fe/fe-core/src/main/java/org/apache/doris/rewrite/BinaryPredicatesDateRule.java

```diff
@@ -0,0 +1,50 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.rewrite;
+
+import org.apache.doris.analysis.Analyzer;
+import org.apache.doris.analysis.BinaryPredicate;
+import org.apache.doris.analysis.CastExpr;
+import org.apache.doris.analysis.Expr;
+import org.apache.doris.analysis.BoolLiteral;
+import org.apache.doris.common.AnalysisException;
+
+/**
+ * Binary predicate date rule try to convert date expression, if date is invalid, it will be
+ * converted into bool literal to avoid to scan all partitions
+ * Examples:
+ * date = "2020-10-32" => FALSE
+ */
+public class BinaryPredicatesDateRule implements ExprRewriteRule {
+    public static ExprRewriteRule INSTANCE = new BinaryPredicatesDateRule();
+
+    @Override
+    public Expr apply(Expr expr, Analyzer analyzer) throws AnalysisException {
+        if (!(expr instanceof BinaryPredicate)) return expr;
+        Expr lchild = expr.getChild(0);
+        if (!lchild.getType().isDateType()) {
+            return expr;
+        }
+        Expr valueExpr = expr.getChild(1);
+        if(valueExpr.getType().isDateType() && valueExpr.isConstant()
+                && valueExpr instanceof CastExpr) {
+            return new BoolLiteral(false);
```

Comment:
You should return null, not false.

Reply:
return null will cause NullPointException:
java.lang.NullPointerException: null
	at org.apache.doris.rewrite.ExprRewriter.applyRuleBottomUp(ExprRewriter.java:82) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.rewrite.ExprRewriter.applyRuleRepeatedly(ExprRewriter.java:71) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.rewrite.ExprRewriter.rewrite(ExprRewriter.java:55) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.analysis.SelectStmt.rewriteExprs(SelectStmt.java:1424) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.qe.StmtExecutor.analyzeAndGenerateQueryPlan(StmtExecutor.java:514) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.qe.StmtExecutor.analyze(StmtExecutor.java:475) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.qe.StmtExecutor.execute(StmtExecutor.java:272) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.qe.StmtExecutor.execute(StmtExecutor.java:243) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:199) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.qe.ConnectProcessor.dispatch(ConnectProcessor.java:337) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.qe.ConnectProcessor.processOnce(ConnectProcessor.java:535) ~[palo-fe.jar:3.4.0]
	at org.apache.doris.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:50) ~[palo-fe.jar:3.4.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_231]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_231]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_231]

Reply:
I mean null Literal.

Reply:
done